File: leetcode-algorithms/653. Two Sum IV - Input is a BS/Solution.cpp
// Given a Binary Search Tree and a target number, return true if there exist two elements in the BST such that their sum is equal to the given target.
//
// Example 1:
// Input:
//     5
//    / \
//   3   6
//  / \   \
// 2   4   7
//
// Target = 9
//
// Output: True
// Example 2:
// Input:
//     5
//    / \
//   3   6
//  / \   \
// 2   4   7
//
// Target = 28
//
// Output: False

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool findTarget(TreeNode* root, int k) {
        vector<int> vec;
        travel(root, vec);

        int a = 0;
        int b = vec.size() - 1;

        while (a < b) {
            int total = vec[a] + vec[b];
            if (total == k) return true;
            else if (total > k) b--;
            else a++;
        }

        return false;
    }

    void travel(TreeNode* root, vector<int>& vec) {
        if (root == nullptr) return;

        travel(root->left, vec);
        vec.push_back(root->val);
        travel(root->right, vec);
    }
};

File: leetcode-algorithms/653. Two Sum IV - Input is a BS/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean findTarget(TreeNode root, int k) {
        visit(root);

        int a = 0;
        int b = vlist.size() - 1;
        while (a < b) {
        	int sum = vlist.get(a) + vlist.get(b);
        	if (sum == k) return true;
        	else if (sum > k) b--;
        	else a++;
        }

        return false;
    }

    private void visit(TreeNode node) {
    	if (node == null) return;

    	visit(node.left);
    	vlist.add(node.val);
    	visit(node.right);
    }

    List<Integer> vlist = new ArrayList<>();
}
